CI Switch to using newer ubuntu images for CI workflows

betatim opened the pull request:
For these workflows it doesn't seem important to pin to a particular version and this way we don't have to periodically update the version.

I got an email from GitHub telling me that `ubuntu-20.04` will be turned off in April and that there will be brown outs before then.

There are also azure pipeline jobs that use 20.04, I've not touched those, but maybe we should do that (in this PR) as well?

---

 The Ubuntu 20.04 runner image will be fully unsupported by April 1, 2025. To raise awareness of the upcoming removal, we will temporarily fail jobs using Ubuntu 20.04. Builds that are scheduled to run during the brownout periods will fail. The brownouts are scheduled for the following dates and times:

    March 4 14:00 UTC – 22:00 UTC
    March 11 13:00 UTC – 21:00 UTC
    March 18 13:00 UTC – 21:00 UTC
    March 25 13:00 UTC – 21:00 UTC

What you need to do

Jobs using the ubuntu-20.04 YAML workflow label should be updated to ubuntu-22.04, ubuntu-24.04 or ubuntu-latest. You can always get up-to-date information on our tools by reading about the software in the runner images repository. Please contact GitHub Support if you run into any problems or need help.

betatim committed:
Switch to using ubuntu-latest image

For these workflows it doesn't seem important to pin to a particular
version and this way we don't have to periodically update the version.

jeremiedbb commented:
> There are also azure pipeline jobs that use 20.04, I've not touched those, but maybe we should do that (in this PR) as well?

+1

jeremiedbb commented:
I don't remember the exact reason for pinning instead of using latest. I think an argument was to be robust to `latest` changing, breaking the CI.

jeremiedbb commented:
On azure pipelines, one of the builds uses ubuntu 20.04 to test on dependencies shipped with this ubuntu. It comes back to the discussion about which oldest Python version we support :)
Bumping ubuntu's version for this job should probably be done alongside bumping our Python min version, but maybe we want to do that in a separate PR.

betatim commented:
> I don't remember the exact reason for pinning instead of using latest. I think an argument was to be robust to `latest` changing, breaking the CI.

I was thinking that for the two label related CI jobs it might be a worthwhile tradeoff to use `latest` as they don't do anything complicated/that relies on the OS version? But it is more a +epsilon from me than a +1

> On azure pipelines, one of the builds uses ubuntu 20.04 to test on dependencies shipped with this ubuntu. It comes back to the discussion about which oldest Python version we support :)

Do you know which job that is? I looked through `azure-pipelines.yml` and the only one I found is https://github.com/scikit-learn/scikit-learn/blob/4ec5f69061a9c37e0f6b9920e296e06c6b4669ac/azure-pipelines.yml#L207 - but this uses conda to setup things :-/ It is testing Python 3.9 but not with what is available from the linux distro. I'm puzzled.

betatim committed:
Pin a concrete version of the OS

Explicit is better than implicit.

betatim committed:
Update Ubuntu version for azure pipelines

Use a newer version of the distro, except for the job  that explicitly
tests things on 20.04.

jeremiedbb commented:
> Do you know which job that is? I looked through azure-pipelines.yml and the only one I found is
> but this uses conda to setup things :-/ It is testing Python 3.9 but not with what is available from the linux distro. I'm puzzled.
> Or is it the job on line 219 which uses pip?

Oh I mixed up 2 jobs :/ You're right, the one checking dependencies shipped with ubuntu has 22.04.

jeremiedbb commented:
> I was thinking that for the two label related CI jobs it might be a worthwhile tradeoff to use latest as they don't do anything complicated/that relies on the OS version?

I agree with that. Were I wasn't sure is for all azure pipeline jobs.

ogrisel reviewed (approved):
LGTM as it is.

jeremiedbb reviewed (approved):
Alright let's merge as is. I still think it's fine to set latest on the gh workflow ones but we can revisit later.
